File: 64-minimum-path-sum/minimum-path-sum.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& points) {
        int m = points.size();
        int n = points[0].size();
        vector<vector<int>> dp(m, vector<int> (n, 0));
        
        dp[0][0] = points[0][0];

        for(int row=0; row<m; row++){
            for(int col=0; col<n; col++){
                if(row == 0 && col == 0) continue;
                dp[row][col] = points[row][col] + min(((row>=1) ? dp[row-1][col]: 1e9), ((col>=1) ? dp[row][col-1] : 1e9));
            }
        }

        return dp[m-1][n-1];

    }

  // Memoization
    int dfs(int row, int col, vector<vector<int>>& points, vector<vector<int>>& dp){

        if(row == 0 && col == 0) return points[0][0];

        else if(row < 0 || col < 0) return 1e9;

        if(dp[row][col] != -1) return dp[row][col];

        int up = dfs(row-1, col, points, dp);
        int left = dfs(row, col-1, points, dp);

        return dp[row][col] = points[row][col] + min(left, up);
    }
};

File: 64-minimum-path-sum/minimum-path-sum.java
class Solution {
    public 
    int minPathSum(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;

        int[][] dp = new int[m][n];

        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                dp[i][j] = -1;
            }
        }

        return dfs(grid, m-1, n-1, dp);
    }

    int dfs(int[][] grid, int m, int n, int[][] dp){
        if(m == 0 && n==0) return grid[0][0];

        if(m < 0 || n < 0) return Integer.MAX_VALUE;

        if(dp[m][n] != -1) return dp[m][n];

        return dp[m][n] = Math.min(dfs(grid, m-1, n, dp), dfs(grid, m, n-1, dp)) + grid[m][n];
    }
}